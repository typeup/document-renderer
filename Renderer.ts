import * as dom from "@typeup/dom"
export type renderFunction = (renderer: Renderer, node: dom.Node) => Promise<string>

export class Renderer {
	private constructor(private variables: { [name: string]: string } = {}) {
	}
	getVariable(name: string): string {
		return this.variables[name]
	}
	setVariable(name: string, value: string) {
		this.variables[name] = value
	}
	async render(node: dom.Node | dom.Node[]): Promise<string> {
		let result: string
		if (Array.isArray(node))
			result = (await Promise.all(node.map(n => this.render(n)))).reduce<string[]>((r, e) => Array.isArray(e) ? [...r, ...e] : [...r, e], []).join("")
		else {
			const render = renderers[node.class]
			result = render ? await render(this, node) : ""
		}
		return result
	}
	scope(): Renderer {
		return new Renderer({ ...this.variables })
	}
	static create(): Renderer {
		return new Renderer({})
	}
}
const renderers: { [className: string]: renderFunction } = {}
export function addRenderer(className: string, render: renderFunction) {
	renderers[className] = render
}
