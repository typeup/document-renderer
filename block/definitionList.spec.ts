import * as dom from "@typeup/dom"
import { Renderer } from "../Renderer"
import "../index"

jest.useFakeTimers()

const renderer = Renderer.create()
describe("Block.DefinitionList", () => {
	it("render", async () => {
		const node = new dom.block.DefinitionList([
			new dom.block.DefinitionTerm([
				new dom.inline.Text("Alpha"),
			], [
				new dom.block.DefinitionData([
					new dom.inline.Text("Term A"),
				]),
				new dom.block.DefinitionData([
					new dom.inline.Text("First Term"),
				]),
			]),
			new dom.block.DefinitionTerm([
				new dom.inline.Text("Beta"),
			], [
				new dom.block.DefinitionData([
					new dom.inline.Text("Term B"),
				]),
				new dom.block.DefinitionData([
					new dom.inline.Text("Second Term"),
				]),
			]),
		])
		expect(await renderer.render(node)).toMatchSnapshot()
	})
})
