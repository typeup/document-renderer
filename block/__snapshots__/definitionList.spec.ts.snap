// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Block.DefinitionList render 1`] = `
"<dl>
	<dt>Alpha</dt>
	<dd>Term A</dd>
	<dd>First Term</dd>
	<dt>Beta</dt>
	<dd>Term B</dd>
	<dd>Second Term</dd>
</dl>"
`;
